### PR TITLE
chore: cleanup mypy config

### LINKED ISSUE #3489: Check adherence with `sp-repo-review` guidelines
There are a lot of interesting things available in the Scientific Python ecosystem, perhaps we should also try using [`sp-repo-review`](https://learn.scientific-python.org/development/guides/repo-review/)?

_Originally posted by @agriyakhetarpal in https://github.com/pybamm-team/PyBaMM/issues/3390#issuecomment-1786984884_

### PR DESCRIPTION
**Pull request by Saransh-cpp**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #3489

## Type of change

Please add a line in the relevant section of [CHANGELOG.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CHANGELOG.md) to document the change (include PR #) - note reverse order of PR #s. If necessary, also add to the list of breaking changes.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (back-end change that speeds up the code)
- [ ] Bug fix (non-breaking change which fixes an issue)

# Key checklist:

- [ ] No style issues: `$ pre-commit run` (or `$ nox -s pre-commit`) (see [CONTRIBUTING.md](https://github.com/pybamm-team/PyBaMM/blob/develop/CONTRIBUTING.md#installing-and-using-pre-commit) for how to set this up to run automatically when committing locally, in just two lines of code)
- [ ] All tests pass: `$ python run-tests.py --all` (or `$ nox -s tests`)
- [ ] The documentation builds: `$ python run-tests.py --doctest` (or `$ nox -s doctests`)

You can run integration tests, unit tests, and doctests together at once, using `$ python run-tests.py --quick` (or `$ nox -s quick`).

## Further checks:

- [ ] Code is commented, particularly in hard-to-understand areas
- [ ] Tests added that prove fix is effective or that feature works

### CODE REVIEW

**Comment by agriyakhetarpal** [comment]
Note: when ready, this PR should close https://github.com/pybamm-team/PyBaMM/issues/3489, since setting up the type checking configurations was the only remaining task on the list.

File: noxfile.py [hunk 1]
```diff
@@ -151,7 +151,7 @@ def run_unit(session):
                 "scikits.odes",
                 external=True,
             )
-            session.install("-e", ".[all,dev,jax,odes]", silent=False)
+            session.install("-e", ".[all,dev,jax]", silent=False)
```
Comment (Saransh-cpp) [inline noxfile.py:154]: Woops
```suggestion
            session.install("-e", ".[all,dev,jax,odes]", silent=False)
```

**Review by agriyakhetarpal — APPROVED** [review]
Thanks, @Saransh-cpp! Question: should we add the `from __future__ import annotations` statement to every file in the source code? We aren't doing it currently, but I do remember that there is a particular tool that does it and lints files for its absence.

**Comment by Saransh-cpp** [comment]
Adding the isort rule blowed up this PR. I'll exclude examples from the isort rule.

**Comment by Saransh-cpp** [comment]
Hi, yes. There are a ton of conflicts and it would just be better to start this PR from scratch. Thanks for the ping!

**Pull request closed without merging**